### PR TITLE
608 - add tck to honour null returned by converter

### LINKED ISSUE #608: Clarify when defaultValue is used on ConfigProperty
**Clarify when defaultValue is used**
`@ConfigProperty(name="my.property", defaultValue="123") int myProp;`
In the [javadoc](https://github.com/eclipse/microprofile-config/blob/master/api/src/main/java/org/eclipse/microprofile/config/inject/ConfigProperty.java#L142) of `ConfigProperty`, it has the following:
The default value if the configured property value does not exist.

What if in a config source, the following config property is defined.
`my.property=abc`.

Normally you will get IllegalArgumentException if the converter throws that. However, what if a converter did not throw an exception, it returns `null`. Will `defaultValue` or `null` be assigned to the variable?

- Option 1: defaultValue is not used and `null` -> myProp. Add a tck to verify this.
Pros: honour the converter. In this way, converter has an option to assign `null`.

- Option 2: defaultValue is used. Javadoc needs to be updated to say: if converter return `null`, this value will be used.
Pros: non null is returned.
Cons: behaviour change. Not possible to return a null. Could be confusing if a property is defined with a value and the defaultValue is still used.

Thoughts?

### PR DESCRIPTION
**Pull request by Emily-Jiang**
Signed-off-by: Emily Jiang <emijiang6@googlemail.com>

Fixes #608

### CODE REVIEW

**Comment by Emily-Jiang** [comment]
@radcortez @jbee please review

File: tck/src/main/java/org/eclipse/microprofile/config/tck/ConvertedNullValueTest.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        WebArchive war = ShrinkWrap
+                .create(WebArchive.class, "ConvertedNullValueTest.war")
+                .addAsLibrary(testJar);
+        return war;
+    }
+
+    @Test
+    public void testDefaultValue() {
+        // This should return null as the converter returns null if the property does not contain :
+        Assert.assertNull(myBean.getPizza());
```
Comment (radcortez) [inline tck/src/main/java/org/eclipse/microprofile/config/tck/ConvertedNullValueTest.java:71]: Hum, shouldn't this fall into the empty rules here: https://github.com/eclipse/microprofile-config/issues/446?

Comment (radcortez) [reply tck/src/main/java/org/eclipse/microprofile/config/tck/ConvertedNullValueTest.java:71]: And in that case, throw an Exception?

Comment (Emily-Jiang) [reply tck/src/main/java/org/eclipse/microprofile/config/tck/ConvertedNullValueTest.java:71]: not really. The empty rull is to throw exception when a `null` passed to converter. This one is different: `cheese` was passed to converter. Converter decides to return null after parsing the input.

Comment (jbee) [reply tck/src/main/java/org/eclipse/microprofile/config/tck/ConvertedNullValueTest.java:71]: Happy to see you confused. This is why I think allowing converters return `null` is no good idea as it is very confusing.

Comment (radcortez) [reply tck/src/main/java/org/eclipse/microprofile/config/tck/ConvertedNullValueTest.java:71]: In my opinion, empty rules must apply to conversion result as well, or we fall again into the same issue where we are going to see inconsistent results when looking up configuration values.

Comment (Emily-Jiang) [reply tck/src/main/java/org/eclipse/microprofile/config/tck/ConvertedNullValueTest.java:71]: I did say this was an experiment :). In this case, I will do a separate PR for Option 2 to see whether it is any better.

**Pull request closed without merging**